[DOCS] Add search and filtering tutorial/quickstart, edit filtering page (#114353)
elastic/elasticsearch: docs/reference/quickstart/full-text-tutorial.asciidoc

We'll use a cooking blog dataset in this tutorial.
Create the `cooking_blog` index and its mapping to get started.

[source,console]
----
PUT /cooking_blog
{
  "mappings": {
    "properties": {
      "title": {
        "type": "text",
        "analyzer": "standard" <1>
      },
      "description": {
        "type": "text"
      },
      "author": {
        "type": "text",
        "fields": {
          "keyword": { <2>
            "type": "keyword",
            "ignore_above": 256
          }
        }
      },
      "date": {
        "type": "date",
        "format": "yyyy-MM-dd"
      },
      "category": {
        "type": "keyword" 
      },
      "tags": {
        "type": "keyword"
      },
      "rating": {
        "type": "float"
      }
    }
  }
}
----
// TESTSETUP
<1> The `standard` analyzer is used by default if an `analyzer` isn't specified.
Note that analyzers are determined per field. Each field can have a different analyzer.
<2> Uses <<multi-fields,multi-fields>> to index `author` as both `text` and `keyword` types. This will enable both full-text search and strict term matches.

[TIP]
====
Full-text search is powered by <<analysis,text analysis>>. 

> Text analysis normalizes and standardizes text data so it can be efficiently stored in an inverted index and searched in near real-time.

Reviewer: Link to docs defining inverted index here?

Author: We don't have this today :sad_panda:.

I would love to have a page about data structures in the [basics section](https://www.elastic.co/guide/en/elasticsearch/reference/current/elasticsearch-intro.html) that summarizes how Elasticsearch stores different data types in optimized structures: inverted index for text, BKD trees for geospatial data, HNSW for vectors etc.

A lot of this core stuff is buried in blogs.

Another disappointing learning doing some of this work is that at some point the Definitive Guide was deprecated but not all of the essential basics were ported over into the docs! 

See https://github.com/elastic/elasticsearch-definitive-guide/blob/master/052_Mapping_Analysis/35_Inverted_index.asciidoc. Need to restore this to the docs ASAP in the right spot.

Analysis happens at both <<analysis-index-search-time,index and search time>>.
====

[discrete]
[[full-text-tutorial-index-data]]
=== Add sample blog posts to your index

Next, index some example blog posts.

[source,console]
----

> POST /cooking_blog/_bulk

Reviewer: Would this make more sense in a notebook?

Author: You mean the whole tutorial? For sure, the plan is to have notebook versions too once have a few of these :)

{"index":{"_id":"4"}}
{"title":"Vegan Chocolate Avocado Mousse","description":"Discover the magic of avocado in this rich, vegan chocolate mousse. Creamy, indulgent, and secretly healthy, it's the perfect guilt-free dessert for chocolate lovers.","author":"Alex Green","date":"2023-05-15","category":"Dessert","tags":["vegan","chocolate","avocado","healthy dessert"],"rating":4.5}
{"index":{"_id":"5"}}
{"title":"Crispy Oven-Fried Chicken","description":"Get that perfect crunch without the deep fryer! This oven-fried chicken recipe delivers crispy, juicy results every time. A healthier take on the classic comfort food.","author":"Maria Rodriguez","date":"2023-05-20","category":"Main Course","tags":["chicken","oven-fried","healthy"],"rating":4.9}
----
// TEST

[discrete]
[[full-text-tutorial-match-query]]
=== Perform basic full-text searches

> 

Reviewer 2: ```suggestion

Full text search consists of performing text queries on different document fields.
These queries will score each document according to how the query text matches the document content.
Different queries provide different ways of looking for the text information and scoring the document.

```

[discrete]
==== `match` query

The <<query-dsl-match-query, `match`>> query is the standard query for full-text AKA "lexical" search.

> Input is analyzed before performing the search by tokenizing text, removing possessives, converting to lowercase, filtering out stop words, and applying stemming.

Reviewer 2: ```suggestion
The query text will be analyzed according to the analyzer configuration specified on each field (or at query time). This provides an opportunity for filtering out stop words, applying stemming to look for different word variations, using synonyms, remove uppercase and lowercase distinctions, etc.
```

Search the `description` field for "fluffy pancakes":

[source,console]
----
GET /cooking_blog/_search
{
  "query": {
    "match": {
      "description": {
        "query": "fluffy pancakes", <1>

>         "analyzer": "standard" <2>

Reviewer 2: Should we move analyzers to a different section? I think that could be initially confusing. Adding analyzers as a secondary step for fine-tuning relevance, applying synonyms, etc could be interesting to have once full-text search is understood.

Author: Indeed it might even be out of scope of this 101 quick start to dig into analyzers.

Author: Just _knowing_ that analyzers are a thing after working thru these examples might be good enough for this section.

<2> {es} defaults to the analyzer defined in the field mapping, so this field is unnecessary here. This is only required if you're using a different analyzer at search time, which is an <<different-analyzers,advanced use case>>.

[discrete]
==== Refine the `match` query

Specify the `and` operator to require both terms in the `description` field.

[source,console]
----
GET /cooking_blog/_search
{
  "query": {
    "match": {
      "description": {
        "query": "fluffy pancakes",
        "operator": "and"
      }
    }
  }
}
----
// TEST[continued]

This stricter query returns no results.

Use the <<query-dsl-minimum-should-match,`minimum_should_match`>> parameter to specify the minimum number of terms a document should have to be included in the search results.

Search the title field to match at least 2 of the 3 terms: "fluffy", "pancakes", or "breakfast".
This is useful for improving relevance while allowing some flexibility.

[source,console]
----
GET /cooking_blog/_search
{
  "query": {
    "match": {
      "title": {
        "query": "fluffy pancakes breakfast",
        "minimum_should_match": 2
      }
    }
  }
}
----
// TEST[continued]

> 

Reviewer 2: Maybe add some links to other full text queries, like match phrase, query string, intervals?

[discrete]
[[full-text-tutorial-multi-match]]
=== Search across multiple fields at once

> A <<query-dsl-multi-match-query,`multi_match`>> query runs the same query against multiple fields.

Reviewer: Worth adding field boosts here. I would personally recommend most text search users use `multi_match` over `match`.

Author: >  I would personally recommend most text search users use multi_match over match.

Cool, what would be the concise explanation for this recommendation?

Author: Handled in https://github.com/elastic/elasticsearch/pull/114353/commits/92e47af3da87e2e4948863f57677789eb3e2176b

[source,console]
----
GET /cooking_blog/_search
{
  "query": {
    "multi_match": {
      "query": "vegetarian curry",
      "fields": ["title", "description", "tags"]
    }
  }
}
----
// TEST[continued]

[discrete]
[[full-text-tutorial-filtering]]
=== Filter and find exact matches

<<filter-context,Filtering>> allows you to narrow down your search results based on exact criteria.
Unlike full-text searches, filters are binary (yes/no) and do not affect the relevance score.

> Filters execute faster than queries.

Reviewer 2: ```suggestion
Filters execute faster than queries as no scoring needs to be calculated for excluded results.
```

[source,console]
----
GET /cooking_blog/_search
{
  "query": {
    "bool": {
      "filter": [
        { "term": { "category.keyword": "Breakfast" } }
      ]
    }
  }
}
----
// TEST[continued]

[discrete]
[[full-text-tutorial-range-query]]
==== Search for posts within a date range

A <<query-dsl-range-query,`range`>> query finds documents that fall within numeric or date ranges.

[source,console]
----
GET /cooking_blog/_search
{
  "query": {
    "range": {
      "date": {
        "gte": "2023-05-01", <1>
        "lte": "2023-05-31" <2>
      }
    }
  }
}
----
// TEST[continued]
<1> Greater than or equal to May 1, 2023.
<2> Less than or equal to May 31, 2023.

[discrete]
[[full-text-tutorial-term-query]]
==== Find exact matches

A <<query-dsl-term-query,`term`>> query searches for an exact term in a field without analyzing it.
Exact, case-sensitive matches on specific terms are often referred to as "keyword" searches.

Search for the author "Maria Rodriguez" in the `author.keyword` field.

[source,console]
----
GET /cooking_blog/_search
{
  "query": {
    "term": {
      "author.keyword": "Maria Rodriguez" <1>
    }
  }
}
----
// TEST[continued]

> <1> The `term` query has zero flexibility. For example, here the queries `maria` or `maria rodriguez` would have zero hits.

Reviewer 2: I'd say this depends on the analyzer used. That would be a good entry point for an analyzers section.

[TIP]
====
Avoid using the `term` query for <<text,`text` fields>> because they are transformed by the analysis process.
====

[discrete]
[[full-text-tutorial-bool-query]]
=== Combine multiple search criteria

> A <<query-dsl-bool-query,`bool`>> query combines multiple queries using boolean logic.

Reviewer: It would be nice to have a final bool query at the end that combines a bunch of stuff - multi match, fuzziness, filters, etc.

Author: Added in https://github.com/elastic/elasticsearch/pull/114353/commits/92e47af3da87e2e4948863f57677789eb3e2176b

      ],
      "filter": [ <2>
        { "term": { "category.keyword": "Breakfast" } },
        { "range": { "rating": { "gte": 4.5 } } }
      ]
    }
  }
}
----
// TEST[continued]
<1> The `must` clause requires the `description` field to match "recipe". This means that only documents containing the word "recipe" in the `description` field will be considered.
<2> The `filter` clauses ensure only documents in the "Breakfast" category with a rating of 4.5 or higher are returned, without affecting the relevance score.

[discrete]
[[full-text-tutorial-fuzzy-query]]
=== Handle typos and variations in search terms

A <<query-dsl-fuzzy-query,`fuzzy` query>> finds terms that are similar to the search term, accounting for typos or slight variations.
Fuzzy queries have no analysis phase.

[source,console]
----
GET /cooking_blog/_search
{
  "query": {
    "fuzzy": {
      "description": {
        "value": "pancaeks", <1>
        "fuzziness": "AUTO" <2>
      }
    }
  }
}
----
// TEST[continued]
<1> The term to search for, allowing for variations.
<2> Automatically determines the appropriate level of fuzziness.

[discrete]
[[full-text-tutorial-prefix-query]]
=== Search by prefixes

> A <<query-dsl-prefix-query,`prefix`>> query enables you to quickly narrow down results to documents that begin with a specified prefix.

Author: 💯 agree that there's a bit too much in here already.

      }
    }
  }
}
----
// TEST[continued]
<1> Because the `title` field is analyzed at index time, it's been processed by the `lowercase` filter. The prefix query isn't analyzed and would return zero hits if you passed "Spicy" here, without setting `case_insensitive` to `true`.

[discrete]
[[full-text-tutorial-query-string]]
=== Perform advanced searches with query strings

A <<query-dsl-query-string-query,`query_string`>> query supports complex queries with a compact syntax.
Use `query_string` for advanced, multi-field searches with complex syntax.

[source,console]
----
GET /cooking_blog/_search
{
  "query": {
    "query_string": {
      "fields": ["title", "description", "tags"], <1>
      "query": "(vegetarian OR vegan) AND (curry OR *fry) AND rating:>=4.5" <2>
    }
  }
}
----
// TEST[continued]
<1> Searches across multiple fields simultaneously
<2> Combines boolean logic, wildcards, and `range` queries:
   - `(vegetarian OR vegan)`: Matches either dietary preference
   - `(curry OR *fry)`: Uses wildcard for various cooking methods
   - `rating:>=4.5`: Filters for highly-rated recipes

[[full-text-tutorial-aggregations]]
[discrete]
=== Perform basic aggregations

> Aggregations provide summary statistics and analytics on your data.

Reviewer 2: ```suggestion
Aggregations provide summary statistics and analytics on your search results.
```

[source, console]
----
GET /cooking_blog/_search
{

>   "size": 0,

Author: ```suggestion
  "size": 0, <1>
```

>   "size": 0,

Author: Out of scope now that we've decided to trim aggregations from this tutorial, but relevant for follow-up  aggregations tutorial

  "aggs": {
    "category_breakdown": {
      "terms": {
        "field": "category.keyword"
      }
    },
    "avg_rating": {
      "avg": {
        "field": "rating"
      }
    }
  }
}
----
// TEST[continued]

> 

Author: ```suggestion
<1> Here you just want the aggregated information on search results, and don't need to return the actual search results themselves.

```

[source,console]
----
GET /cooking_blog/_search
{
  "query": {
    "match": {
      "title": {
        "query": "fluffy pancakes breakfast",
        "minimum_should_match": 2
      }
    }
  }
}
----
// TEST[continued]

[discrete]
[[full-text-tutorial-multi-match]]
=== Search across multiple fields at once

A <<query-dsl-multi-match-query,`multi_match`>> query runs the same query against multiple fields.

[source,console]
----
GET /cooking_blog/_search
{
  "query": {
    "multi_match": {
      "query": "vegetarian curry",
      "fields": ["title", "description", "tags"]
    }
  }
}
----
// TEST[continued]

[discrete]
[[full-text-tutorial-filtering]]
=== Filter and find exact matches

<<filter-context,Filtering>> allows you to narrow down your search results based on exact criteria.
Unlike full-text searches, filters are binary (yes/no) and do not affect the relevance score.
Filters execute faster than queries as no scoring needs to be calculated for excluded results.

This `bool` query will return only blog posts in the "Breakfast" category.

[source,console]
----
GET /cooking_blog/_search
{
  "query": {
    "bool": {
      "filter": [

>         { "term": { "category.keyword": "Breakfast" } }

Reviewer 3: Could be good to call out the usage of the `keyword` multi-field here

----
GET /cooking_blog/_search
{
  "query": {
    "prefix": { 
      "title": {
        "value": "Spicy",
        "case_insensitive": true <1>
      }
    }
  }
}
----
// TEST[continued]
<1> Because the `title` field is analyzed at index time, it's been processed by the `lowercase` filter. The prefix query isn't analyzed and would return zero hits if you passed "Spicy" here, without setting `case_insensitive` to `true`.

[discrete]
[[full-text-tutorial-query-string]]
=== Perform advanced searches with query strings

A <<query-dsl-query-string-query,`query_string`>> query supports complex queries with a compact syntax.
Use `query_string` for advanced, multi-field searches with complex syntax.

[source,console]
----
GET /cooking_blog/_search
{
  "query": {
    "query_string": {
      "fields": ["title", "description", "tags"], <1>
      "query": "(vegetarian OR vegan) AND (curry OR *fry) AND rating:>=4.5" <2>
    }
  }
}
----
// TEST[continued]
<1> Searches across multiple fields simultaneously
<2> Combines boolean logic, wildcards, and `range` queries:
   - `(vegetarian OR vegan)`: Matches either dietary preference
   - `(curry OR *fry)`: Uses wildcard for various cooking methods

>    - `rating:>=4.5`: Filters for highly-rated recipes

Reviewer 3: There's a lot in this example, we should expand the description

Author: might be one for more advanced follow-up

[[full-text-tutorial-aggregations]]
[discrete]
=== Perform basic aggregations

> Aggregations provide summary statistics and analytics on your search results.

Reviewer 3: Can we link to a page with more detailed info on aggregations here?

Author: out of scope now, we'll have an aggregations heavy tutorial methinks
